File: ios/TeeTimeCaddie/TeeTimeCaddie/Source/auth/FirebaseAuthService.swift
//
//  FirebaseAuthService.swift
//  TeeTimeCaddie
//
//  Platform (iOS) implementation of the shared AuthService interface, backed by the native
//  Firebase Auth SDK. Created by the app and passed into the SDK during initialization, so the
//  shared code carries no Firebase dependency. Native Firebase errors are classified into the
//  shared AuthErrors taxonomy here.
//

import Foundation
import TeeTimeCaddieKit
import FirebaseAuth

class FirebaseAuthService: AuthService {

    private let auth: Auth

    init(useEmulator: Bool) {
        auth = Auth.auth()
        if useEmulator {
            auth.useEmulator(withHost: Self.emulatorHost, port: Self.emulatorPort)
        }
    }

    var currentUserId: String? { auth.currentUser?.uid }
    var currentUserDisplayName: String? { auth.currentUser?.displayName }

    // Suspend requirements are exposed by SKIE with a `__` prefix (see FirebaseFirestoreClient).
    func __signIn(email: String, password: String) async throws -> AuthResult {
        do {
            let result = try await auth.signIn(withEmail: email, password: password)
            return AuthResultSuccess(userId: result.user.uid)
        } catch {
            return AuthResultFailure(error: TeeTimeCaddieKit.AuthErrors.invalidCredentials)
        }
    }

    func __register(email: String, password: String) async throws -> AuthResult {
        do {
            let result = try await auth.createUser(withEmail: email, password: password)
            return AuthResultSuccess(userId: result.user.uid)
        } catch let error as NSError {
            return AuthResultFailure(error: mapRegisterError(error))
        }
    }

    func __updateDisplayName(name: String) async throws {
        guard let user = auth.currentUser else { return }
        let request = user.createProfileChangeRequest()
        request.displayName = name
        try await request.commitChanges()
    }

    func __refreshToken() async throws -> KotlinBoolean {
        guard let user = auth.currentUser else { return KotlinBoolean(bool: false) }
        do {
            _ = try await user.getIDTokenResult(forcingRefresh: true)
            return KotlinBoolean(bool: true)
        } catch {
            return KotlinBoolean(bool: false)
        }
    }

    func signOut() {
        try? auth.signOut()
    }

    func observeAuthState(onChange: @escaping (String?) -> Void) -> TeeTimeCaddieKit.Cancellable {
        let handle = auth.addStateDidChangeListener { _, user in
            onChange(user?.uid)
        }
        return AuthStateCancellable(auth: auth, handle: handle)
    }

    private func mapRegisterError(_ error: NSError) -> TeeTimeCaddieKit.AuthErrors {
        switch error.code {
        case AuthErrorCode.emailAlreadyInUse.rawValue: return TeeTimeCaddieKit.AuthErrors.userExists
        case AuthErrorCode.invalidEmail.rawValue: return TeeTimeCaddieKit.AuthErrors.invalidEmail
        case AuthErrorCode.weakPassword.rawValue: return TeeTimeCaddieKit.AuthErrors.weakPassword
        default: return TeeTimeCaddieKit.AuthErrors.regDefault
        }
    }

    private static let emulatorHost = "127.0.0.1"
    private static let emulatorPort = 9099
}

private class AuthStateCancellable: TeeTimeCaddieKit.Cancellable {
    private let auth: Auth
    private let handle: AuthStateDidChangeListenerHandle
    init(auth: Auth, handle: AuthStateDidChangeListenerHandle) {
        self.auth = auth
        self.handle = handle
    }
    func cancel() { auth.removeStateDidChangeListener(handle) }
}

File: android/app/src/main/java/net/bradball/teetimecaddie/android/auth/FirebaseAuthService.kt
package net.bradball.teetimecaddie.android.auth

import com.google.android.gms.tasks.Task
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthException
import com.google.firebase.auth.FirebaseAuthInvalidCredentialsException
import com.google.firebase.auth.FirebaseAuthUserCollisionException
import com.google.firebase.auth.UserProfileChangeRequest
import kotlinx.coroutines.suspendCancellableCoroutine
import net.bradball.teetimecaddie.core.models.storage.Cancellable
import net.bradball.teetimecaddie.features.auth.AuthErrors
import net.bradball.teetimecaddie.features.auth.AuthResult
import net.bradball.teetimecaddie.features.auth.AuthService
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Android [AuthService] implementation backed by the native Firebase Auth SDK.
 *
 * This is the platform-provided binding that replaces the GitLive Firebase Auth library. It maps
 * native Firebase auth error types to the shared [AuthErrors] taxonomy.
 */
class FirebaseAuthService(useEmulator: Boolean) : AuthService {

    private val auth: FirebaseAuth = FirebaseAuth.getInstance().apply {
        if (useEmulator) useEmulator(EMULATOR_HOST, EMULATOR_PORT)
    }

    override val currentUserId: String? get() = auth.currentUser?.uid
    override val currentUserDisplayName: String? get() = auth.currentUser?.displayName

    override suspend fun signIn(email: String, password: String): AuthResult {
        return try {
            val result = auth.signInWithEmailAndPassword(email, password).awaitResult()
            AuthResult.Success(result.user?.uid ?: auth.currentUser?.uid.orEmpty())
        } catch (ex: Exception) {
            AuthResult.Failure(AuthErrors.INVALID_CREDENTIALS)
        }
    }

    override suspend fun register(email: String, password: String): AuthResult {
        return try {
            val result = auth.createUserWithEmailAndPassword(email, password).awaitResult()
            val uid = result.user?.uid ?: return AuthResult.Failure(AuthErrors.REG_DEFAULT)
            AuthResult.Success(uid)
        } catch (ex: FirebaseAuthUserCollisionException) {
            AuthResult.Failure(AuthErrors.USER_EXISTS)
        } catch (ex: FirebaseAuthInvalidCredentialsException) {
            AuthResult.Failure(AuthErrors.INVALID_EMAIL)
        } catch (ex: FirebaseAuthException) {
            AuthResult.Failure(AuthErrors.REG_DEFAULT)
        }
        // Non-Firebase exceptions (e.g. network) propagate and are classified as UNKNOWN by the repository.
    }

    override suspend fun updateDisplayName(name: String) {
        val user = auth.currentUser ?: return
        val request = UserProfileChangeRequest.Builder().setDisplayName(name).build()
        user.updateProfile(request).awaitResult()
    }

    override suspend fun refreshToken(): Boolean {
        val user = auth.currentUser ?: return false
        return try {
            user.getIdToken(true).awaitResult()
            true
        } catch (ex: Exception) {
            false
        }
    }

    override fun signOut() {
        auth.signOut()
    }

    override fun observeAuthState(onChange: (String?) -> Unit): Cancellable {
        val listener = FirebaseAuth.AuthStateListener { firebaseAuth ->
            onChange(firebaseAuth.currentUser?.uid)
        }
        auth.addAuthStateListener(listener)
        return object : Cancellable {
            override fun cancel() = auth.removeAuthStateListener(listener)
        }
    }

    companion object {
        private const val EMULATOR_HOST = "10.0.2.2"
        private const val EMULATOR_PORT = 9099
    }
}

private suspend fun <T> Task<T>.awaitResult(): T = suspendCancellableCoroutine { continuation ->
    addOnSuccessListener { result -> if (continuation.isActive) continuation.resume(result) }
    addOnFailureListener { error -> if (continuation.isActive) continuation.resumeWithException(error) }
}
